### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by SashaPylypyuk**
https://sashapylypyuk.github.io/react_add-todo-form/

### CODE REVIEW

File: src/App.js [hunk 1]
```diff
@@ -1,17 +1,15 @@
...
 function App() {
   return (
     <div className="App">
       <h1>Add todo form</h1>
-
-      <p>
-        <span>Users: </span>
-        {users.length}
-      </p>
+      <TodoList users={users} todos={todos} />
```
Comment (irenhh) [inline src/App.js]: create one array with todo and corresponding user in each element

File: src/components/TodoList/TodoList.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  addTodo = (todo) => {
+    this.setState(prevState => ({
+      todos: [...prevState.todos, todo],
+    }));
+  }
+
+  render() {
+    return (
+      <>
+        <NewTodo users={this.state.users} addTodo={this.addTodo} />
```
Comment (irenhh) [inline src/components/TodoList/TodoList.js]: destructure this.state

File: src/components/TodoList/TodoList.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        <table className="table">
+          <thead>
+            <tr className="table__head">
+              <th>Id</th>
+              <th>Title</th>
+              <th>User Id</th>
+            </tr>
+          </thead>
+          {this.state.todos.map(todo => (
+            <tr className="table__row">
```
Comment (irenhh) [inline src/components/TodoList/TodoList.js]: where is key attribute?

File: src/components/NewTodo/NewTodo.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    errorUser: false,
+  }
+
+  onSubmit = (e) => {
+    e.preventDefault();
+
+    const { userId, id, title } = this.state;
+
+    if (title === '') {
+      this.setState({
```
Comment (irenhh) [inline src/components/NewTodo/NewTodo.js]: use setState once for one method

Comment (irenhh) [reply src/components/NewTodo/NewTodo.js]: write all the code inside setState callback

File: src/components/NewTodo/NewTodo.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      return;
+    }
+
+    const todo = {
+      userId, id, title,
+    };
+
+    this.props.addTodo(todo);
+    this.setState({
+      id: id + 1,
```
Comment (irenhh) [inline src/components/NewTodo/NewTodo.js]: don't mutate state

File: src/components/NewTodo/NewTodo.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  render() {
+    const { users } = this.state;
+
+    return (
+      <>
+        <form onSubmit={this.onSubmit} className="form">
+          <input
+            value={this.state.title}
```
Comment (irenhh) [inline src/components/NewTodo/NewTodo.js]: destructure

File: src/App.js (hunk 1 above)
Comment (SashaPylypyuk) [reply src/App.js]: I don't understand why we need one array for todos and users. In "TodoList" we push users in "NewTodo" and must show them in the list, in array where we have todos and users it's hard make. I think we don't need one array

File: src/components/NewTodo/NewTodo.js (hunk 4 above)
Comment (SashaPylypyuk) [reply src/components/NewTodo/NewTodo.js]: i can't validate properties in setState's callback

File: src/components/NewTodo/NewTodo.js (hunk 5 above)
Comment (SashaPylypyuk) [reply src/components/NewTodo/NewTodo.js]: Image 1

File: src/components/NewTodo/NewTodo.js (hunk 4 above)
Comment (irenhh) [reply src/components/NewTodo/NewTodo.js]: why? use prevState

File: src/components/NewTodo/NewTodo.js (hunk 5 above)
Comment (irenhh) [reply src/components/NewTodo/NewTodo.js]: just prevState.id + 1

File: src/components/NewTodo/NewTodo.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    if (errorUser || errorTitle) {
+      return;
+    }
+
+    const todo = {
+      userId, id, title,
+    };
+
+    this.props.addTodo(todo);
+    this.setState(prevState => ({
```
Comment (irenhh) [inline src/components/NewTodo/NewTodo.js]: you're still using two setStates in one method

Comment (SashaPylypyuk) [reply src/components/NewTodo/NewTodo.js]: Так в цьому методі неможливо використовувати 1 setState, спочатку йде валідація потім добавлення в масив і потім обнулення через setState, чесно я не розумыю як тут використовувати 1 setState

Comment (irenhh) [reply src/components/NewTodo/NewTodo.js]:
```
this.setState((prevState) => {
  all your validations
  when you want to return new state object just write
  return {
    new state object
  }
})
```

File: src/components/NewTodo/NewTodo.js [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        return {
+          errorUser: true,
+        };
+      }
+
+      const todo = {
+        userId, id, title,
+      };
+
+      this.props.addTodo(todo);
```
Comment (irenhh) [inline src/components/NewTodo/NewTodo.js]: don't call addTodo in setState

Comment (irenhh) [reply src/components/NewTodo/NewTodo.js]: make a separate validation function and return some object with errors or just one value for example and then check it in this method and return setState with errors on addTodo in `else` statement

**Pull request closed without merging**